### ISSUE TITLE
Two broadcasts of the same source shape share one index expression, and the second one wins

### ISSUE THREAD

**Issue by benkelaya** [bledden/triton-msl#9]
**Author:** Hocine Benkelaya (with NeuroBrix)

## The point that decides

The same expression, compiled twice. Alone it is correct; in a kernel that
also broadcasts a *differently shaped* tensor from the same source shape, it
is not.

```python
flip = core.reshape(
    core.broadcast_to(core.arange(0, 2)[None, :, None], [1, 2, 4]), [8])
```

A `1x2x1` tensor broadcast to `1x2x4` holds four `0`s then four `1`s, so the
reshape to `[8]` is `[0,0,0,0,1,1,1,1]`.

| kernel | result |
|---|---|
| the expression alone | `[0, 0, 0, 0, 1, 1, 1, 1]` — correct |
| the same expression beside a `1x2x1 -> 4x2x1` broadcast | `[0, 1, 0, 1, 0, 1, 0, 1]` |

`[0,1,0,1,0,1,0,1]` is the *source's* two elements read cyclically: a
broadcast that was never materialised.

The TTIR is right in both cases and identical for the part in question:

```mlir
%flip   = tt.make_range {end = 2 : i32, start = 0 : i32} : tensor<2xi32>
%flip_2 = tt.expand_dims %flip   {axis = 0 : i32} : tensor<2xi32>   -> tensor<1x2xi32>
%flip_3 = tt.expand_dims %flip_2 {axis = 2 : i32} : tensor<1x2xi32> -> tensor<1x2x1xi32>
%flip_4 = tt.broadcast   %flip_3 : tensor<1x2x1xi32> -> tensor<1x2x4xi32>
%flip_5 = tt.reshape     %flip_4 : tensor<1x2x4xi32> -> tensor<8xi32>
```

## What is actually emitted

Alone — the index expression is the right one for a `1x2x4` target, whose
middle axis has stride 4:

```c
kernel void flip_alone(...) {
        uint idx_0 = ((lid / 4u) % 2u);
        o_flip[(0 + lid)] = idx_0;
}
```

Beside a second broadcast whose source is also `1x2x1` but whose target is
`4x2x1` (middle-axis stride 1):

```c
kernel void flip_beside(...) {
        float val_3 = static_cast<float>(xv[(0 + lid)]);
        uint idx_4 = (lid % 2u);      // <- flip, for a 1x2x4 target
        uint idx_5 = (lid % 2u);      // <- the mask, for a 4x2x1 target
        bool mask_6 = idx_5 == 0;
        ...
        o_flip[(0 + lid)] = idx_4;
}
```

`idx_4` and `idx_5` are the same expression. `idx_5` is correct for its own
target; `idx_4` should be `(lid / 4u) % 2u`, as the first kernel shows it can
be. The middle-axis index appears to be derived from the source shape alone —
both sources are `tensor<1x2x1>` — without the broadcast target's shape, so
whichever target is emitted last decides for both.

Nothing warns. The kernel compiles, runs, and returns an array of the right
shape and plausible contents.

## Reproducer

```python
import numpy as np, torch, triton, triton.language as tl
from triton.language import core

N, ND, STAGE, I = 8, 3, 2, 2

@triton.jit
def flip_alone(o, xv, N: tl.constexpr, ND: tl.constexpr, STAGE: tl.constexpr):
    cols = tl.arange(0, N)
    x = tl.load(xv + cols).to(tl.float32)
    n_outer: tl.constexpr = N >> ND
    fshape: tl.constexpr = [n_outer * 2 ** (ND - 1 - STAGE), 2, 2 ** STAGE]
    flip = core.reshape(
        core.broadcast_to(core.arange(0, 2)[None, :, None], fshape), x.shape)
    tl.store(o + cols, flip.to(tl.int32))

@triton.jit
def flip_beside(o, o_l, o_r, xv, N: tl.constexpr, ND: tl.constexpr,
                STAGE: tl.constexpr, I: tl.constexpr):
    cols = tl.arange(0, N)
    x = tl.load(xv + cols).to(tl.float32)
    n_outer: tl.constexpr = N >> ND
    fshape: tl.constexpr = [n_outer * 2 ** (ND - 1 - STAGE), 2, 2 ** STAGE]
    flip = core.reshape(
        core.broadcast_to(core.arange(0, 2)[None, :, None], fshape), x.shape)
    shape: tl.constexpr = [n_outer * 2 ** I, 2, 2 ** (ND - I - 1)]
    y = core.reshape(x, shape)
    m = core.arange(0, 2)[None, :, None]
    left = core.reshape(core.broadcast_to(
        tl.sum(core.where(m == 0, y, tl.zeros_like(y)), 1)[:, None, :], shape), x.shape)
    right = core.reshape(core.broadcast_to(
        tl.sum(core.where(m == 1, y, tl.zeros_like(y)), 1)[:, None, :], shape), x.shape)
    tl.store(o + cols, flip.to(tl.int32))
    tl.store(o_l + cols, left)
    tl.store(o_r + cols, right)

x = torch.tensor([3.,1.,4.,1.,5.,9.,2.,6.], device="mps")
f = torch.zeros(N, dtype=torch.int32, device="mps")
l = torch.zeros(N, dtype=torch.float32, device="mps")
r = torch.zeros(N, dtype=torch.float32, device="mps")

flip_alone[(1,)](f, x, N=N, ND=ND, STAGE=STAGE); torch.mps.synchronize()
print("alone :", f.cpu().tolist())
f.zero_()
flip_beside[(1,)](f, l, r, x, N=N, ND=ND, STAGE=STAGE, I=I); torch.mps.synchronize()
print("beside:", f.cpu().tolist())
print("left  :", l.cpu().tolist())
print("right :", r.cpu().tolist())
```

```
alone : [0, 0, 0, 0, 1, 1, 1, 1]
beside: [0, 1, 0, 1, 0, 1, 0, 1]
left  : [3.0, 3.0, 4.0, 4.0, 5.0, 5.0, 2.0, 2.0]
right : [1.0, 1.0, 1.0, 1.0, 9.0, 9.0, 6.0, 6.0]
```

## It is an interference, not one broken construct

Both constructs are correct on their own, and the reproducer shows it in the
same run:

* removing the second broadcast makes `flip` correct — first kernel;
* the second broadcast's own result is correct while `flip` is wrong — `left`
  and `right` above are exactly the two lanes of the pairs `(3,1) (4,1) (5,9)
  (2,6)`.

So neither expression is at fault by itself. What fails is the pair.

## Why it matters

This is a silent wrong rather than a refusal, and it lands in the middle of a
bitonic sort. The `flip` vector is what tells a bitonic stage which direction
each block sorts in; with the alternating pattern replaced by a cyclic one,
compare-and-swap exchanges the wrong pairs and the XOR swap duplicates values
instead of permuting them. A sort of eight floats then returns a tile whose
multiset is not the input's, with no error and no warning.

The public `tl.sort` does not go through this path — its `_compare_and_swap`
was rewritten and no longer builds `flip` by reshaping a broadcast — so the
in-tree tests do not reach it. Any out-of-tree kernel still using the older
reshape-based form does.

## A correction to an earlier analysis of mine

I first attributed this to the compare-and-swap machinery itself, and drafted
an issue saying so. Writing a standalone reproducer refuted it: `tl.sort` is
correct at every size here, and a second hypothesis — that a 3-D reduce
silently drops a pre-reduce elementwise op — is refuted by the `left`/`right`
values above, which are exact. Neither claim survived measurement, and neither
was filed. The reproducer in this issue is what is left after both were
removed.

Worth noting in passing, since it may be the same walk: written with only
*one* lane extracted instead of two, the kernel above is **refused** at
compile time with

```
Refusing to emit silently-wrong output: 3D reduce with a pre-reduce
elementwise op ... the generic 3D-reduce path mis-computes this shape.
```

The two-lane form compiles. I have not established whether that guard and this
index collapse share a cause; I mention it only because a guard that fires on
the smaller case and not the larger one is the kind of asymmetry worth a look
from someone who knows the pass.

## Environment

Apple M4 Pro, macOS 26.6.2 (Darwin 25.6.0), Python 3.12.0, triton 3.7.0,
Metal backend.

**Comment by bledden** [comment]
Thank you for isolating the interference and correcting the earlier hypotheses. I agree that this report isn't evidence that public `tl.sort` is broken or that the two-lane reduction dropped its input arithmetic.

In 0.3.0, the broadcast index handling accounts for each range's target shape instead of letting the neighboring broadcast supply the wrong stride. The reported case and derived comparison/select cases have regression pins, including the `[0,0,0,0,1,1,1,1]` pattern beside the differently shaped broadcast. That addresses the corruption mechanism in the older reshape-based sort spelling, not just whether it compiles.

Your one-lane versus two-lane observation also deserves an answer. The single 3-D reduction can hit the specialized template's pre-reduce-operation refusal, while multiple reductions bypass that detector and take the generic route. The broadcast repair does not remove that separate restriction or establish that both forms have identical support. I am treating the one-lane refusal as a remaining capability limitation, not counting it as fixed by the collision test.

If your larger sorting example still fails, a synthetic input and the exact kernel/launch would let me check both ordering and preservation of the input values. A passing public `tl.sort` test wouldn't substitute for testing that older spelling.

**Comment by benkelaya** [comment]
Thanks, agreed on both points: the broadcast repair, and the one-lane refusal as a separate capability limit. If the larger sort still fails on 0.3.0, we'll send a synthetic input with the exact kernel and launch.
